src/main/java/de/dennisguse/opentracks/viewmodels/MarkerEditViewModel.java

```diff
@@ -0,0 +1,173 @@
+        if (marker.hasPhoto()) {
+            if (!marker.getPhotoURI().equals(photoOriginalUri)) {
+                deletePhoto(marker.getPhotoURI());
+            }
+            marker.setPhotoUrl(null);
+            marker.setName(name);
+            marker.setCategory(category);
+            marker.setDescription(description);
+            markerData.postValue(marker);
+        }
+    }
+
+    public void onNewCameraPhoto(@NonNull Uri photoUri, String name, String category, String description) {
+        Marker marker =  getMarkerOrThrowException();
+        marker.setPhotoUrl(photoUri.toString());
+        marker.setName(name);
+        marker.setCategory(category);
+        marker.setDescription(description);
+        markerData.postValue(marker);
+    }
+
+    public void onNewGalleryPhoto(@NonNull Uri srcUri, String name, String category, String description) {
+        Marker marker =  getMarkerOrThrowException();
+
+        try (ParcelFileDescriptor parcelFd = getApplication().getContentResolver().openFileDescriptor(srcUri, "r")) {
+            FileDescriptor srcFd = parcelFd.getFileDescriptor();
+            File dstFile = new File(FileUtils.getImageUrl(getApplication(), marker.getTrackId()));
+            FileUtils.copy(srcFd, dstFile);
+
+            Uri photoUri = FileUtils.getUriForFile(getApplication(), dstFile);
+            marker.setPhotoUrl(photoUri.toString());
+            marker.setName(name);
+            marker.setCategory(category);
+            marker.setDescription(description);
+            markerData.postValue(marker);
+        } catch(Exception e) {
+            Log.e(TAG, e.getMessage());
+            Toast.makeText(getApplication(), R.string.marker_add_canceled, Toast.LENGTH_LONG).show();
+        }
+    }
+
+    private void onAddDone(@NonNull Marker marker, String name, String category, String description) {
+        trackRecordingServiceConnection.addMarker(getApplication(), name, category, description, marker.hasPhoto() ? marker.getPhotoURI().toString() : null);
+    }
+
+    private void onSaveDone(@NonNull Marker marker, String name, String category, String description) {
+        marker.setName(name);
+        marker.setCategory(category);
+        marker.setDescription(description);
+        new ContentProviderUtils(getApplication()).updateMarker(getApplication(), marker);
+
+        if (photoOriginalUri != null && (!marker.hasPhoto() || !photoOriginalUri.equals(marker.getPhotoURI()))) {
+            deletePhoto(photoOriginalUri);
+        }
+    }
+
+    public void onDone(String name, String category, String description) {
+        Marker marker = getMarkerOrThrowException();
+        if (isNewMarker) {
+            onAddDone(marker, name, category, description);
+        } else {
+            onSaveDone(marker, name, category, description);
+        }
+    }
+
+    public void onCancel() {
+        Marker marker =  getMarkerOrThrowException();
+        if (isNewMarker) {
```

Comment:
Somehow relatively complicated (or it is too late).
Is it possible that multiple if's are valid?

Reply:
Simplified...